hazardous/survtrace/_preprocessor.py

```diff
@@ -0,0 +1,93 @@
+import numpy as np
+import pandas as pd
+from sklearn.base import BaseEstimator, TransformerMixin
+from sklearn.compose import make_column_transformer
+from sklearn.preprocessing import OrdinalEncoder, StandardScaler
+
+
+class FeaturePreprocessor(TransformerMixin, BaseEstimator):
+    def __init__(self, categ_cols, num_cols):
+        self.categ_cols = categ_cols
+        self.num_cols = num_cols
+
+    def fit(self, X, y=None):
+        self.col_transformer_ = make_column_transformer(
+            (OrdinalEncoder(), self.categ_cols),
+            (StandardScaler(), self.num_cols),
+            remainder="drop",
+            verbose_feature_names_out=False,
+        )
+        self.col_transformer_.set_output(transform="pandas")
+        self.col_transformer_.fit(X)
+
+        # Encode categorical values from different columns separately
+        # e.g. "blue" in column 1 is represented by a different token
+        # than "blue" in column 2.
+        transformers = self.col_transformer_.transformers_
+        categories = transformers[0][1].categories_
+        vocab_size = [0, *[len(categs) for categs in categories[:-1]]]
+        self.vocab_size_ = np.cumsum(vocab_size)
+
+        return self
+
+    def transform(self, X, y=None):
+        X_t = self.col_transformer_.transform(X)
+        X_t[self.categ_cols] += self.vocab_size_
+
+        return X_t
+
+
+class TargetPreprocessor(TransformerMixin, BaseEstimator):
+    def __init__(self, horizons):
+        self.horizons = horizons
+
+    def fit(self, y):
+        # Check strictly increasing
+        if not np.all(self.horizons[:-1] < self.horizons[1:]):
+            raise ValueError(
+                f"'horizon' must be strictly increasing, got {self.horizons=!r}"
+            )
+
+        # Defining cuts
+        mask_first_event = y["event"] == 1
```

Comment:
This follows the SurvTRACE implementation, but we can probably use an "any event" mask instead.